File: src/quarantine/quarantineStore.ts
import { randomBytes } from "node:crypto";
import { sha256Hex } from "../canonicalJson.js";
import { HttpError } from "../httpError.js";
import type { BankId, QuarantineKind, ReviewReason } from "../types.js";
import {
  createEncryptedQuarantineEnvelope,
  decodePublicKey,
  type DecryptedQuarantineObject,
} from "./envelopeCrypto.js";
import {
  InMemorySlidingWindowRateLimiter,
  type QuarantineRateLimiter,
  type RateLimitSession,
} from "./rateLimiter.js";
import type {
  NewQuarantineItem,
  QuarantineCapacityLimits,
  QuarantineRepository,
} from "./repository.js";

export interface QuarantineInput {
  timestamp: string;
  kind: QuarantineKind;
  reason: ReviewReason;
  writerId?: string;
  source?: string;
  sourceBank?: BankId;
  sourceMemoryId?: string;
  sourceContentSha256?: string;
  dedupeKey?: string;
  payload: unknown;
}

export interface QuarantineResult {
  quarantine_id: string;
  sha256: string;
}

export interface QuarantineStore {
  put(input: QuarantineInput): Promise<QuarantineResult>;
}

export interface QuarantineStoreLimits {
  maxItemBytes: number;
  maxPendingItems: number;
  maxEncryptedBytes: number;
  rateLimitMax: number;
  rateLimitWindowMs: number;
  rateLimitGlobalMax: number;
  requarantineOpsMax: number;
}

export const DEFAULT_QUARANTINE_LIMITS: QuarantineStoreLimits = {
  maxItemBytes: 1_048_576,
  maxPendingItems: 1_000,
  maxEncryptedBytes: 104_857_600,
  rateLimitMax: 30,
  rateLimitWindowMs: 60_000,
  rateLimitGlobalMax: 300,
  requarantineOpsMax: 1_000,
};

const GLOBAL_WRITES_BUCKET = "quarantine-writes";
const REQUARANTINE_OPS_BUCKET = "quarantine-requarantine-ops";
const AUTH_AUDIT_WRITES_BUCKET = "quarantine-writes:auth-audit";
const AUTH_AUDIT_REQUARANTINE_OPS_BUCKET =
  "quarantine-requarantine-ops:auth-audit";
const UNKNOWN_WRITER_BUCKET = "unknown-writer";

export class EncryptedDatabaseQuarantineStore implements QuarantineStore {
  private readonly publicKey: string;
  private readonly capacity: QuarantineCapacityLimits;
  private readonly rateLimiter: QuarantineRateLimiter;

  constructor(
    publicKey: string,
    readonly repository: QuarantineRepository,
    private readonly limits: QuarantineStoreLimits = DEFAULT_QUARANTINE_LIMITS,
    rateLimiter?: QuarantineRateLimiter,
  ) {
    this.publicKey = decodePublicKey(publicKey);
    this.capacity = {
      maxPendingItems: limits.maxPendingItems,
      maxEncryptedBytes: limits.maxEncryptedBytes,
    };
    this.rateLimiter = rateLimiter ?? new InMemorySlidingWindowRateLimiter();
  }

  async put(input: QuarantineInput): Promise<QuarantineResult> {
    const quarantineId = await this.resolveQuarantineId(input);
    const encrypted = this.encrypt(input, quarantineId);
    const encryptedBytes = Buffer.byteLength(JSON.stringify(encrypted));
    if (encryptedBytes > this.limits.maxItemBytes) {
      throw new HttpError(
        413,
        "quarantine_item_too_large",
        "encrypted quarantine item exceeds configured size limit",
      );
    }

    return this.rateLimiter.withIdentityLock(
      quarantineId,
      async (rateLimitSession) => {
        const knownIdentity = await this.isKnownIdentity(input, quarantineId);
        await this.chargeRateQuota(input, knownIdentity, rateLimitSession);

        const item = this.buildItem(input, quarantineId, encrypted);
        if (input.kind === "recalled_memory") {
          await this.repository.upsertRecalledMemory(item, this.capacity);
        } else if (input.kind === "security_event") {
          await this.repository.upsertSecurityEvent(item, this.capacity);
        } else if (item.dedupe_key) {
          await this.repository.upsertRequestItem(item, this.capacity);
        } else {
          await this.repository.insert(item, this.capacity);
        }

        return { quarantine_id: quarantineId, sha256: encrypted.sha256 };
      },
    );
  }

  private encrypt(input: QuarantineInput, quarantineId: string) {
    const decrypted: DecryptedQuarantineObject = {
      quarantine_id: quarantineId,
      created_at: input.timestamp,
      reason: input.reason,
      ...(input.writerId === undefined ? {} : { writer_id: input.writerId }),
      ...(input.source === undefined ? {} : { source: input.source }),
      payload: input.payload,
    };
    return createEncryptedQuarantineEnvelope(decrypted, this.publicKey);
  }

  private buildItem(
    input: QuarantineInput,
    quarantineId: string,
    encrypted: ReturnType<typeof createEncryptedQuarantineEnvelope>,
  ): NewQuarantineItem {
    return {
      quarantine_id: quarantineId,
      created_at: input.timestamp,
      updated_at: input.timestamp,
      kind: input.kind,
      reason: input.reason,
      ...(input.writerId === undefined ? {} : { writer_id: input.writerId }),
      ...(input.source === undefined ? {} : { source: input.source }),
      ...(input.sourceBank === undefined
        ? {}
        : { source_bank: input.sourceBank }),
      ...(input.sourceMemoryId === undefined
        ? {}
        : { source_memory_id: input.sourceMemoryId }),
      ...(input.sourceContentSha256 === undefined
        ? {}
        : { source_content_sha256: input.sourceContentSha256 }),
      ...(input.dedupeKey === undefined ? {} : { dedupe_key: input.dedupeKey }),
      sha256: encrypted.sha256,
      encrypted,
      status: "pending",
      postpone_count: 0,
    };
  }

  private async chargeRateQuota(
    input: QuarantineInput,
    knownIdentity: boolean,
    rateLimitSession: RateLimitSession,
  ): Promise<void> {
    const windowMs = this.limits.rateLimitWindowMs;
    // Auth-failure audits use dedicated buckets so unauthenticated probing
    // can never share budget with security quarantines or their refreshes.
    const authAudit = input.reason === "auth_failed";
    if (knownIdentity) {
      await rateLimitSession.consume(
        authAudit
          ? AUTH_AUDIT_REQUARANTINE_OPS_BUCKET
          : REQUARANTINE_OPS_BUCKET,
        { max: this.limits.requarantineOpsMax, windowMs },
      );
      return;
    }
    if (this.limits.rateLimitMax <= 0 || (await this.capacityExhausted())) {
      return;
    }
    if (authAudit) {
      await rateLimitSession.consume(AUTH_AUDIT_WRITES_BUCKET, {
        max: this.limits.rateLimitGlobalMax,
        windowMs,
      });
      return;
    }

    const writer = input.writerId ?? UNKNOWN_WRITER_BUCKET;
    await rateLimitSession.consumeMany([
      {
        key: `${GLOBAL_WRITES_BUCKET}:writer:${writer}`,
        rule: { max: this.limits.rateLimitMax, windowMs },
      },
      {
        key: GLOBAL_WRITES_BUCKET,
        rule: { max: this.limits.rateLimitGlobalMax, windowMs },
      },
    ]);
  }

  private async isKnownIdentity(
    input: QuarantineInput,
    quarantineId: string,
  ): Promise<boolean> {
    if (input.kind === "security_event" && input.dedupeKey) {
      return (await this.repository.get(quarantineId)) !== null;
    }
    if (
      (input.kind === "retain_request" || input.kind === "recall_request") &&
      input.dedupeKey
    ) {
      return (await this.repository.get(quarantineId)) !== null;
    }
    if (
      input.kind === "recalled_memory" &&
      input.sourceBank !== undefined &&
      input.sourceMemoryId !== undefined
    ) {
      return (
        (await this.repository.findMemoryState(
          input.sourceBank,
          input.sourceMemoryId,
        )) !== null
      );
    }
    return false;
  }

  private async capacityExhausted(): Promise<boolean> {
    const stats = await this.repository.stats();
    return (
      stats.pending_items + stats.postponed_items >=
        this.capacity.maxPendingItems ||
      stats.encrypted_bytes >= this.capacity.maxEncryptedBytes
    );
  }

  private async resolveQuarantineId(input: QuarantineInput): Promise<string> {
    if (input.kind === "security_event" && input.dedupeKey) {
      const digest = sha256Hex(input.dedupeKey);
      return `q_security${digest.slice(0, 48)}_${digest.slice(48)}`;
    }
    if (
      (input.kind === "retain_request" || input.kind === "recall_request") &&
      input.dedupeKey
    ) {
      const digest = sha256Hex(input.dedupeKey);
      return `q_request${digest.slice(0, 48)}_${digest.slice(48)}`;
    }
    if (
      input.kind === "recalled_memory" &&
      input.sourceBank !== undefined &&
      input.sourceMemoryId !== undefined
    ) {
      const digest = sha256Hex(`${input.sourceBank}:${input.sourceMemoryId}`);
      return `q_memory${digest.slice(0, 48)}_${digest.slice(48)}`;
    }
    return `q_${input.timestamp.replace(/[^0-9A-Za-z]/g, "")}_${randomBytes(8).toString("hex")}`;
  }
}

File: memory_router/quarantine_store.py
from __future__ import annotations

import re
import secrets
from dataclasses import dataclass
from datetime import UTC, datetime, timedelta
from typing import Any, cast

from .canonical import sha256_hex
from .dedupe import request_family_identity
from .envelope import create_envelope, decode_public_key, estimate_envelope_size
from .errors import HttpError
from .repository import Capacity, QuarantineRepository


@dataclass(frozen=True, slots=True)
class QuarantineLimits:
    max_item_bytes: int = 1_048_576
    max_pending_items: int = 1_000
    max_pending_items_per_writer: int = 50
    max_encrypted_bytes: int = 104_857_600
    rate_limit_max: int = 30
    rate_limit_window_ms: int = 60_000
    rate_limit_global_max: int = 300
    distinct_family_limit_max: int = 10
    requarantine_ops_max: int = 1_000
    item_ttl_days: int = 0


class QuarantineStore:
    def __init__(
        self,
        public_key: str,
        repository: QuarantineRepository,
        limits: QuarantineLimits,
        rate_limiter: Any,
    ) -> None:
        key = decode_public_key(public_key)
        self.public_key = public_key
        self.public_key_bytes = key.key_size // 8
        self.repository = repository
        self.limits = limits
        self.rate_limiter = rate_limiter
        self.capacity = Capacity(
            limits.max_pending_items, _effective_writer_limit(limits), limits.max_encrypted_bytes
        )

    async def put(self, input_: dict[str, Any]) -> dict[str, str]:
        quarantine_id = self._resolve_id(input_)
        self._assert_item_size(input_, quarantine_id)
        existing_for_charge = await self.repository.get(quarantine_id)
        known = await self._known_identity(input_, existing_for_charge is not None)
        await self._charge(input_, known, self.rate_limiter)

        async def operation(_session: Any) -> dict[str, str]:
            existing = await self.repository.get(quarantine_id)
            if (
                input_["kind"] in {"retain_request", "recall_request"}
                and input_.get("dedupeKey")
                and existing
                and existing["status"] not in {"pending", "postponed"}
            ):
                raise HttpError(
                    409,
                    "quarantine_request_in_review",
                    "matching quarantine request is already being reviewed",
                )
            if existing and existing["status"] == "review_in_progress":
                raise HttpError(
                    409,
                    "quarantine_item_in_review",
                    "matching quarantine item is already being reviewed",
                )

            encrypted = self._encrypt(input_, quarantine_id)
            item = self._build_item(input_, quarantine_id, encrypted)
            mode = (
                "memory"
                if input_["kind"] == "recalled_memory"
                else "request"
                if input_["kind"] in {"retain_request", "recall_request"} and item.get("dedupe_key")
                else "id"
            )
            capacity = self.capacity
            if input_["kind"] == "security_event":
                capacity = Capacity(capacity.max_pending_items, 0, capacity.max_encrypted_bytes)
            await self.repository.store(item, capacity, mode=mode, at=input_["timestamp"])
            return {"quarantine_id": quarantine_id, "sha256": str(encrypted["sha256"])}

        return cast(
            dict[str, str], await self.rate_limiter.with_identity_lock(quarantine_id, operation)
        )

    def _decrypted(self, input_: dict[str, Any], quarantine_id: str) -> dict[str, Any]:
        decrypted: dict[str, Any] = {
            "quarantine_id": quarantine_id,
            "created_at": input_["timestamp"],
            "reason": input_["reason"],
            "payload": input_["payload"],
        }
        if input_.get("writerId") is not None:
            decrypted["writer_id"] = input_["writerId"]
        if input_.get("source") is not None:
            decrypted["source"] = input_["source"]
        return decrypted

    def _assert_item_size(self, input_: dict[str, Any], quarantine_id: str) -> None:
        decrypted = self._decrypted(input_, quarantine_id)
        encrypted_bytes = estimate_envelope_size(decrypted, self.public_key_bytes)
        if encrypted_bytes > self.limits.max_item_bytes:
            raise HttpError(
                413,
                "quarantine_item_too_large",
                "encrypted quarantine item exceeds configured size limit",
            )

    def _encrypt(self, input_: dict[str, Any], quarantine_id: str) -> dict[str, Any]:
        return create_envelope(self._decrypted(input_, quarantine_id), self.public_key)

    def _build_item(
        self, input_: dict[str, Any], quarantine_id: str, encrypted: dict[str, Any]
    ) -> dict[str, Any]:
        item: dict[str, Any] = {
            "quarantine_id": quarantine_id,
            "created_at": input_["timestamp"],
            "updated_at": input_["timestamp"],
            "kind": input_["kind"],
            "reason": input_["reason"],
            "writer_id": input_.get("writerId"),
            "source": input_.get("source"),
            "source_bank": input_.get("sourceBank"),
            "source_memory_id": input_.get("sourceMemoryId"),
            "source_content_sha256": input_.get("sourceContentSha256"),
            "dedupe_key": input_.get("dedupeKey"),
            "sha256": encrypted["sha256"],
            "encrypted": encrypted,
            "status": "pending",
            "postpone_count": 0,
            "requarantine_count": 0,
        }
        if self.limits.item_ttl_days > 0:
            try:
                created = datetime.fromisoformat(input_["timestamp"].replace("Z", "+00:00"))
            except ValueError as exc:
                raise HttpError(
                    400,
                    "invalid_quarantine_timestamp",
                    "quarantine timestamp must be an ISO timestamp",
                ) from exc
            item["expires_at"] = (
                (created + timedelta(days=self.limits.item_ttl_days))
                .astimezone(UTC)
                .isoformat(timespec="milliseconds")
                .replace("+00:00", "Z")
            )
        return item

    async def _known_identity(self, input_: dict[str, Any], item_exists: bool) -> bool:
        if input_.get("dedupeKey") and input_["kind"] in {
            "security_event",
            "retain_request",
            "recall_request",
        }:
            return item_exists
        if (
            input_["kind"] == "recalled_memory"
            and input_.get("sourceBank") is not None
            and input_.get("sourceMemoryId") is not None
        ):
            return (
                await self.repository.find_memory_state(
                    input_["sourceBank"], input_["sourceMemoryId"]
                )
                is not None
            )
        return False

    async def _charge(self, input_: dict[str, Any], known: bool, session: Any) -> None:
        window = self.limits.rate_limit_window_ms
        auth_audit = input_["reason"] == "auth_failed"
        if known:
            key = (
                "quarantine-requarantine-ops:auth-audit"
                if auth_audit
                else "quarantine-requarantine-ops"
            )
            await session.consume_many([(key, self.limits.requarantine_ops_max, window)])
            return
        if self.limits.rate_limit_max <= 0:
            return
        if auth_audit:
            await session.consume_many(
                [("quarantine-writes:auth-audit", self.limits.rate_limit_global_max, window)]
            )
            return
        writer = (
            "unknown-writer"
            if input_["reason"] == "unknown_writer"
            else input_.get("writerId") or "unknown-writer"
        )
        family = request_family_identity(
            input_["kind"], input_["reason"], input_.get("writerId"), input_["payload"]
        )
        identities: list[tuple[str, str, int, int]] = []
        if family:
            identities.append(
                (
                    f"quarantine-request-family:{family[0]}",
                    family[1],
                    self.limits.distinct_family_limit_max,
                    window,
                )
            )
        await session.consume_many_distinct(
            [
                (f"quarantine-writes:writer:{writer}", self.limits.rate_limit_max, window),
                ("quarantine-writes", self.limits.rate_limit_global_max, window),
            ],
            identities,
        )

    def _resolve_id(self, input_: dict[str, Any]) -> str:
        if input_["kind"] == "security_event" and input_.get("dedupeKey"):
            digest = sha256_hex(input_["dedupeKey"])
            return f"q_security{digest[:48]}_{digest[48:]}"
        if input_["kind"] in {"retain_request", "recall_request"} and input_.get("dedupeKey"):
            digest = sha256_hex(input_["dedupeKey"])
            return f"q_request{digest[:48]}_{digest[48:]}"
        if (
            input_["kind"] == "recalled_memory"
            and input_.get("sourceBank") is not None
            and input_.get("sourceMemoryId") is not None
        ):
            digest = sha256_hex(f"{input_['sourceBank']}:{input_['sourceMemoryId']}")
            return f"q_memory{digest[:48]}_{digest[48:]}"
        stamp = re.sub(r"[^0-9A-Za-z]", "", input_["timestamp"])
        return f"q_{stamp}_{secrets.token_hex(8)}"


def _effective_writer_limit(limits: QuarantineLimits) -> int:
    if limits.max_pending_items_per_writer <= 0:
        return 0
    return min(
        limits.max_pending_items_per_writer,
        max(0, limits.max_pending_items - 1),
        max(0, (limits.max_encrypted_bytes - 1) // limits.max_item_bytes),
    )
